internal/querycoordv2/checkers/segment_checker.go

```diff
@@ -0,0 +1,242 @@
+		replicas := c.meta.ReplicaManager.GetByCollection(cid)
+		for _, r := range replicas {
+			tasks = append(tasks, c.checkReplica(ctx, r)...)
+		}
+	}
+
+	// find already released segments which are not contained in target
+	segments := c.dist.SegmentDistManager.GetAll()
+	released := utils.FilterReleased(segments, collectionIDs)
+	tasks = append(tasks, c.createSegmentReduceTasks(ctx, released, -1, querypb.DataScope_All)...)
+	return tasks
+}
+
+func (c *SegmentChecker) checkReplica(ctx context.Context, replica *meta.Replica) []task.Task {
+	ret := make([]task.Task, 0)
+	targets := c.targetMgr.GetSegmentsByCollection(replica.CollectionID)
+	dists := c.getSegmentsDist(replica)
+
+	// compare with targets to find the lack and redundancy of segments
+	lacks, redundancies := diffSegments(targets, dists)
+	tasks := c.createSegmentLoadTasks(ctx, lacks, replica)
+	ret = append(ret, tasks...)
+
+	tasks = c.createSegmentReduceTasks(ctx, redundancies, replica.GetID(), querypb.DataScope_All)
+	ret = append(ret, tasks...)
+
+	// compare inner dists to find repeated loaded segments
+	redundancies = findRepeatedSegments(dists)
+	redundancies = c.filterExistedOnLeader(replica, redundancies)
+	tasks = c.createSegmentReduceTasks(ctx, redundancies, replica.GetID(), querypb.DataScope_All)
+	ret = append(ret, tasks...)
+
+	// release redundant growing segments
+	leaderReduancies := c.findNeedReleasedGrowingSegments(replica)
+	redundancies = make([]*meta.Segment, 0)
+	for leaderID, segmentIDs := range leaderReduancies {
+		segments := packSegments(segmentIDs, leaderID, replica.GetCollectionID())
+		redundancies = append(redundancies, segments...)
+	}
+	tasks = c.createSegmentReduceTasks(ctx, redundancies, replica.GetID(), querypb.DataScope_Streaming)
+	ret = append(ret, tasks...)
+
+	return ret
+}
+
+func (c *SegmentChecker) getSegmentsDist(replica *meta.Replica) []*meta.Segment {
+	ret := make([]*meta.Segment, 0)
+	for _, node := range replica.Nodes.Collect() {
+		ret = append(ret, c.dist.SegmentDistManager.GetByCollectionAndNode(replica.CollectionID, node)...)
+	}
+	return ret
+}
+
+func diffSegments(targets []*datapb.SegmentInfo, dists []*meta.Segment) (lacks []*datapb.SegmentInfo, redundancies []*meta.Segment) {
+	distMap := make(map[int64]struct{})
+	targetMap := make(map[int64]struct{})
+	for _, s := range targets {
+		targetMap[s.GetID()] = struct{}{}
+	}
+	for _, s := range dists {
+		distMap[s.GetID()] = struct{}{}
+		if _, ok := targetMap[s.GetID()]; !ok {
+			redundancies = append(redundancies, s)
+		}
+	}
+	for _, s := range targets {
+		if _, ok := distMap[s.GetID()]; !ok {
+			lacks = append(lacks, s)
+		}
+	}
+	return
+}
+
+func findRepeatedSegments(dists []*meta.Segment) []*meta.Segment {
+	ret := make([]*meta.Segment, 0)
+	versions := make(map[int64]*meta.Segment)
+	for _, s := range dists {
+		maxVer, ok := versions[s.GetID()]
+		if !ok {
+			versions[s.GetID()] = s
+			continue
+		}
+		if maxVer.Version <= s.Version {
+			ret = append(ret, maxVer)
```

Comment:
Why append MaxVer into ret repeatedly,  the operation is confusing and lack of more test cases.